services/endpointsrv/service.go

```diff
@@ -438,17 +438,71 @@ func (e *EndpointService) EndpointCheckStatus(ctx context.Context, checkID strin
 
 	out := make([]types.EndpointCheckStep, len(steps))
 	for idx, step := range steps {
+		status, conclusion := StepStatusAndConclusion(step)
 		out[idx] = types.EndpointCheckStep{
-			StepID:    step.ID,
-			EvalID:    step.EvalID,
-			Input:     []byte(step.Input.String),
-			Output:    []byte(step.Output.String),
-			Assertion: step.Assertion.String,
+			StepID:     step.ID,
+			EvalID:     step.EvalID,
+			Input:      []byte(step.Input.String),
+			Output:     []byte(step.Output.String),
+			Assertion:  step.Assertion.String,
+			Status:     status,
+			Conclusion: conclusion,
 		}
 	}
 
+	status, conclusion := CheckStatusAndConclusion(out)
 	return types.EndpointCheck{
-		CheckID: checkID,
-		Steps:   out,
+		CheckID:    checkID,
+		Steps:      out,
+		Status:     status,
+		Conclusion: conclusion,
 	}, nil
 }
+
+// StepStatusAndConclusion infers the status of a step based on the contents of the database.
+func StepStatusAndConclusion(step db.UnweaveEndpointCheckStep) (types.CheckStatus, *types.CheckConclusion) {
+	if !step.Output.Valid {
+		return types.CheckPending, nil
+	}
+
+	if !step.Assertion.Valid {
+		return types.CheckInProgress, nil
+	}
+
+	switch step.Assertion.String {
+	case "":
+		return types.CheckInProgress, nil
+	case "success":
+		return types.CheckCompleted, &types.CheckSuccess
+	case "failure":
+		return types.CheckCompleted, &types.CheckFailure
+	default:
+		return types.CheckCompleted, &types.CheckError
+	}
+}
+
+// CheckStatusAndConclusion infers the status of a check based on the status of the steps.
+func CheckStatusAndConclusion(steps []types.EndpointCheckStep) (types.CheckStatus, *types.CheckConclusion) {
+	conclusions := map[string]struct{}{}
```

Comment:
**suggestion:**: make the map `map[types.CheckConclusion]bool` and most of the rest of the code is easier to read. 

We won't have to do `Conclusion.String()` everywhere we key the map.

I _know_ it's neat to use the zero-bytes of a `struct{}{}` , but if we use bool, and call the map `conculsionsHas[types.CheckConculstion]bool` then we can make the checks really sweeeet! 😍 

```go
if conculsionsHas[types.CheckError] {
    return types.CheckCompleted, &types.CheckError
}
```